module/zfs/mmp.c

```diff
@@ -0,0 +1,548 @@
+	abd_zero(ub_abd, VDEV_UBERBLOCK_SIZE(vd));
+	abd_copy_from_buf(ub_abd, ub, sizeof (uberblock_t));
+
+	mutex_exit(&mmp->mmp_io_lock);
+
+	offset = VDEV_UBERBLOCK_OFFSET(vd, VDEV_FIRST_MMP_BLOCK(vd) +
+	    spa_get_random(MMP_BLOCKS_PER_LABEL));
+
+	vdev_label_write(zio, vd, spa_get_random(VDEV_LABELS),
+	    ub_abd, offset, VDEV_UBERBLOCK_SIZE(vd), mmp_write_done, mmp,
+	    flags | ZIO_FLAG_DONT_PROPAGATE);
+
+	zio_nowait(zio);
+}
+
+static void
+mmp_thread(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	uint64_t last_zfs_mmp_interval = zfs_mmp_interval;
+	boolean_t last_spa_suspended = spa_suspended(spa);
+	callb_cpr_t cpr;
+
+	mmp_thread_enter(mmp, &cpr);
+
+	mmp->mmp_last_write = gethrtime();
+
+	for (;;) {
+		/* for stable values within an iteration */
+		uint64_t mmp_fail_intervals = zfs_mmp_fail_intervals;
+		uint64_t mmp_interval = zfs_mmp_interval;
+		boolean_t suspended = spa_suspended(spa);
+		hrtime_t start, next_time;
+		hrtime_t max_fail_ns = (MSEC2NSEC(mmp_interval) *
+		    mmp_fail_intervals);
+
+		if (mmp->mmp_thread_exiting)
+			goto cleanup_and_exit;
+
+		start = gethrtime();
+		if (mmp_interval) {
+			next_time = start + MSEC2NSEC(mmp_interval) /
+			    vdev_count_leaves(dp->dp_spa);
+		} else {
+			next_time = start + MSEC2NSEC(MMP_DEFAULT_INTERVAL);
+		}
+
+		/*
+		 * When MMP goes off => on, or spa goes suspended =>
+		 * !suspended, we know no writes occurred recently.  We
+		 * update mmp_last_write to give us some time to try.
+		 */
+		if ((!last_zfs_mmp_interval && mmp_interval) ||
+		    (last_spa_suspended && !suspended)) {
+			dprintf("mmp_thread transtion: pool %s last interval "
+			    "%lu interval %lu last suspended %d suspended %d\n",
+			    spa->spa_name, last_zfs_mmp_interval, mmp_interval,
+			    last_spa_suspended, suspended);
+			mutex_enter(&mmp->mmp_io_lock);
+			mmp->mmp_last_write = gethrtime();
+			mutex_exit(&mmp->mmp_io_lock);
+		} else if (last_zfs_mmp_interval && !mmp_interval) {
+			mutex_enter(&mmp->mmp_io_lock);
+			mmp->mmp_delay = 0;
+			mutex_exit(&mmp->mmp_io_lock);
+		}
+		last_zfs_mmp_interval = mmp_interval;
+		last_spa_suspended = suspended;
+
+		/*
+		 * Check after the transition check above.
+		 */
+		if (!suspended && mmp_fail_intervals && mmp_interval &&
+		    (start - mmp->mmp_last_write) > max_fail_ns) {
+			dprintf("mmp suspending pool: pool %s "
+			    "zfs_mmp_interval %lu zfs_mmp_fail_intervals %lu "
+			    "mmp_last_write %llu now %llu\n", spa->spa_name,
+			    mmp_interval, mmp_fail_intervals,
+			    mmp->mmp_last_write, start);
+			zio_suspend(spa, NULL);
+		}
+
+		if (zfs_mmp_interval) {
+			spa_config_enter(spa, SCL_STATE, FTAG, RW_READER);
+			mmp_write_uberblock(dp);
+			spa_config_exit(spa, SCL_STATE, FTAG);
+		}
+
+		if (gethrtime() >= next_time)
+			continue;
+
+		CALLB_CPR_SAFE_BEGIN(&cpr);
+		(void) cv_timedwait_hires(&mmp->mmp_thread_cv,
+		    &mmp->mmp_thread_lock, next_time, NANOSEC,
```

Comment:
I changed the resolution to 1, so that `cv_timedwait_hires()` doesn't attempt to quantize the time passed into `cv_timedwait_hires_common()`.  

I decided against using the resolution as a rate-limiting mechanism (by putting in 10 msec and using the roundup flag) as we discussed.  A system designer with a disk count so high that it makes MMP a performance problem can adjust `zfs_mmp_interval` appropriately, and will likely be tuning other zfs parameters anyway.